src/swap/defi/thorchainDa.ts

```diff
@@ -0,0 +1,547 @@
+import { toFixed } from 'biggystring'
+import {
+  asArray,
+  asBoolean,
+  asEither,
+  asNull,
+  asObject,
+  asOptional,
+  asString,
+  asUnknown
+} from 'cleaners'
+import {
+  EdgeCorePluginOptions,
+  EdgeCurrencyWallet,
+  EdgeSpendInfo,
+  EdgeSwapInfo,
+  EdgeSwapPlugin,
+  EdgeSwapQuote,
+  EdgeSwapRequest,
+  EdgeTransaction,
+  JsonObject,
+  SwapCurrencyError
+} from 'edge-core-js/types'
+import { ethers } from 'ethers'
+
+// import nodeFetch from 'node-fetch'
+import {
+  checkInvalidCodes,
+  getTokenId,
+  makeQueryParams,
+  makeSwapPluginQuote
+} from '../../swap-helpers'
+import { fetchInfo, fetchWaterfall, promiseWithTimeout } from '../../util/utils'
+import { abiMap } from './abi/abiMap'
+import {
+  asExchangeInfo,
+  asInboundAddresses,
+  EVM_CURRENCY_CODES,
+  EXCHANGE_INFO_UPDATE_FREQ_MS,
+  EXPIRATION_MS,
+  getApprovalData,
+  getEvmTokenData,
+  getGasLimit,
+  INVALID_CURRENCY_CODES,
+  MAINNET_CODE_TRANSCRIPTION,
+  MIDGARD_SERVERS_DEFAULT
+} from './thorchain'
+
+const pluginId = 'thorchainda'
+const swapInfo: EdgeSwapInfo = {
+  pluginId,
+  displayName: 'Thorchain DEX Aggregator',
+  supportEmail: 'support@edge.app'
+}
+
+const asInitOptions = asObject({
+  affiliateFeeBasis: asOptional(asString, '50'),
+  thorname: asOptional(asString, 'ej')
+})
+
+interface ThorSwapQuoteParams {
+  sellAsset: string
+  buyAsset: string
+  sellAmount: string
+  slippage: string // 5 = 5%
+  recipientAddress: string
+  senderAddress?: string
+  affiliateAddress: string
+  affiliateBasisPoints: string // '50' => 0.5%
+}
+
+const asCalldata = asObject({
+  tcMemo: asOptional(asString),
+  memo: asOptional(asString)
+})
+
+const asThorSwapQuoteResponse = asObject({
+  contract: asEither(asString, asNull),
+  contractMethod: asEither(asString, asNull),
+  contractInfo: asOptional(asString),
+  complete: asBoolean,
+  path: asString,
+  providers: asArray(asString),
+  calldata: asUnknown,
+  expectedOutput: asString,
+  expectedOutputMaxSlippage: asString,
+  expectedOutputUSD: asString,
+  expectedOutputMaxSlippageUSD: asString,
+  deadline: asOptional(asString)
+})
+
+const DA_VOLATILITY_SPREAD_DEFAULT = 0.03
+const THORSWAP_DEFAULT_SERVERS = [
+  'https://aggregator-prod-aulilvmdlq-uc.a.run.app'
+]
+
+type ExchangeInfo = ReturnType<typeof asExchangeInfo>
+
+let exchangeInfo: ExchangeInfo | undefined
+let exchangeInfoLastUpdate: number = 0
+
+const tokenProxyMap: { [currencyPluginId: string]: string } = {
+  ethereum: '0xf892fef9da200d9e84c9b0647ecff0f34633abe8',
+  avalanche: '0x69ba883af416ff5501d54d5e27a1f497fbd97156'
+}
+
+export function makeThorchainDaPlugin(
+  opts: EdgeCorePluginOptions
+): EdgeSwapPlugin {
+  const { io, log } = opts
+  const { fetch } = io
+  // const fetch = nodeFetch ?? io.fetch
```

Comment:
Leftover line.

Reply:
io.fetch is broken for some API endpoints when running in node. This was added as a convenience so unit tests could work. Either need to troubleshoot io.fetch in node or leave this is for testing convenience (and as a note).

Reply:
Example error I get with `io.fetch` under node that works with `node-fetch`

`"{\"status_code\":1,\"message\":\"Unknown API endpoint /v2/thorchain/inbound_addresses\"}"`